### PR TITLE
Use concepts in algorithm and sequtils #7996

### PR DESCRIPTION
**Pull request by b3liever**

### CODE REVIEW

**Comment by b3liever** [comment]
Fails bootstrapping: reorder.nim(232, 30) Error: undeclared field: 'pnode'
the code calls algorithm.sortedbyit

File: lib/std/baseconcepts.nim [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+  Indexable*[T] = concept c
+    var i: int
+    c[i] is T
+    c[i] is var T
+    var val: T
+    c[i] = val
+    c.len is int
+
+#   StringLike* = concept c of Indexable[char]
```
Comment (mratsim) [inline lib/std/baseconcepts.nim]: This should be removed from this PR

**Comment by Araq** [comment]
Hmm, but interfaces should be minimal and ``low(MyEnum)`` does not produce int, so the common lookup array ``array[MyEnum, string]`` would not be Indexable.

**Comment by bluenote10** [comment]
Indeed, but since the concept uses `var i: int` for the index specification it is already incompatible with `array[MyEnum, string]` without adding low/high. What about modifying it as follows, this seems to work:

```nimrod

type
  Indexable*[T] = concept c
    # somehow infer the index type
    type Index = type(c.low)
    c.high is Index
    var i: Index
    c[i] is T
    c[i] is var T
    var val: T
    c[i] = val
    c.len is int

type
  MyEnum = enum
    A, B

proc test[T](s: Indexable[T]) =
  for i in s.low .. s.high:
    echo s[i]

let t: array[MyEnum, int] = [1, 2]
test(t)
```

**Comment by Araq** [comment]
@bluenote10 I really like it.

**Comment by krux02** [comment]
At the moment concepts are not stable enough to migrate the standard library to use them. But apart from that, it would be ok.

**Comment by b3liever** [comment]
Agreed. Close?

**Comment by bluenote10** [comment]
My impression was that making this change wasn't in the distant future after the recent improvements of concepts. What exactly is blocking it?

**Comment by krux02** [comment]
concepts have exponential compile time explosions for some corner cases. Until that problem isn't solved, I am strictly against concept usage in the standard library anywhere.

**Pull request closed without merging**